### PR TITLE
build: make makefile work on windows

### PR DESCRIPTION
**Pull request by dundargoc**
Using powershell as the default windows shell as using cmd alters $PATH
in a way that makes building neovim fail (powershell prioritizes visual
studio tools which is arguably more correct).

This was tested with gnu make for windows, which can be installed with
e.g. scoop. It does not work with nmake and it is extremely unlikely we
want to add nmake support as the makefile is merely supposed to be
syntactic sugar for the most common case. For similar reasons, the only
supported generator is ninja.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,3 +1,17 @@
...
+MKDIR := mkdir -force
+TOUCH := new-item -force
+CMAKE := cmake
+CMAKE_GENERATOR := Ninja
+else
+MKDIR := mkdir -p
+TOUCH := touch
+CMAKE := $(shell (command -v cmake3 || echo cmake))
+CMAKE_GENERATOR ?= $(shell (command -v ninja > /dev/null 2>&1 && echo "Ninja") || echo "Unix Makefiles")
+endif
```
Comment (lewis6991) [inline Makefile:13]:
```suggestion
ifeq ($(OS),Windows_NT)
  SHELL := powershell.exe
  .SHELLFLAGS := -NoProfile -Command
  MKDIR := mkdir -force
  TOUCH := new-item -force
  CMAKE := cmake
  CMAKE_GENERATOR := Ninja
else
  MKDIR := mkdir -p
  TOUCH := touch
  CMAKE := $(shell (command -v cmake3 || echo cmake))
  CMAKE_GENERATOR ?= $(shell (command -v ninja > /dev/null 2>&1 && echo "Ninja") || echo "Unix Makefiles")
endif
```

Note: spaces (not tabs) are used here.

**Review by lewis6991 — APPROVED** [review]

**Review by justinmk — APPROVED** [review]
cool

**Comment by dundargoc** [comment]
Todo:

fix clean

Emphasize it's only for gnu make and not nmake/borland/solariswhatevermake and only ninja generator. I can't be bothered to accommodate every random usecase, we have cmake for that.

File: Makefile [hunk 2]
```diff
@@ -1,3 +1,17 @@
+ifeq ($(OS),Windows_NT)
+  SHELL := powershell.exe
+  .SHELLFLAGS := -NoProfile -Command
```
Comment (justinmk) [inline Makefile:3]:
```suggestion
  .SHELLFLAGS := -NoLogo -NoProfile -Command
```

**Comment by justinmk** [comment]
> Maybe fix performance? It's slow af for some reason. Low priority


Maybe try `-ExecutionPolicy Bypass` ?

File: Makefile (hunk 2 above)
Comment (dundargoc) [reply Makefile:3]: I don't get it. It doesn't seem to make any difference. The copyright text isn't shown anyway when using the makefile. I'll add it if you still want it.

**Comment by dundargoc** [comment]
> Maybe try -ExecutionPolicy Bypass ?

Benchmarking with hyperfine made no difference. I will experiment with cmd to see to see if that helps.

**Comment by dundargoc** [comment]
I give up on trying to improve performance for now. Powershell is just destined to be slow it seems. We can iterate on it in the future if we need to improve it.

**Comment by clason** [comment]
Nice clean-up in any case, even for other platforms 👍

File: Makefile (hunk 2 above)
Comment (justinmk) [reply Makefile]: more of a "best practice" (avoiding unnecessary output could avoid accidental problems?)

Comment (dundargoc) [reply Makefile]: Sure, added.

**Pull request merged**